Feature: move partitions

zhiqiangxu opened the pull request.

zhiqiangxu committed:
init

zhiqiangxu committed:
avoid recompute current_deadline

zhiqiangxu committed:
treat empty bitfield as all

zhiqiangxu committed:
rm useless quote

zhiqiangxu committed:
add verify

zhiqiangxu committed:
combine option1 & option2

zhiqiangxu committed:
fix

steven004 commented on actors/miner/src/deadlines.rs:
```diff
@@ -128,6 +128,26 @@ pub fn deadline_available_for_compaction(
 
+// the distance between from_deadline and to_deadline clockwise in deadline unit.
+fn deadline_distance(policy: &Policy, from_deadline: u64, to_deadline: u64) -> u64 {
+    if to_deadline > from_deadline {
+        to_deadline - from_deadline
+    } else {
+        policy.wpost_period_deadlines - from_deadline + to_deadline
```
```suggestion
    if to_deadline >= from_deadline {
        to_deadline - from_deadline
    } else {
        policy.wpost_period_deadlines - from_deadline + to_deadline
```

The distance should be 0~47.  
And, this change makes the function deadline_available_for_move more reasonable. It will automatically return false if currDeadline=fromDeadline

zhiqiangxu replied:
Good catch, fixed!

steven004 commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+                ));
+            }
+
+            if !deadline_available_for_move(
+                policy,
+                params.from_deadline,
+                params.to_deadline,
+                current_deadline.index,
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to a deadline which is further from current deadline"
+                ));
+            }
```
Could we move this before the transaction handling? 
I would think it is better if it can fail faster.

zhiqiangxu replied:
Oops, seems not , because `current_deadline` depends on `state`, which is only available in the transaction.

steven004 commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+                    })?;
+            }
+
+            if !deadline_available_for_compaction(
+                policy,
+                current_deadline.period_start,
+                params.to_deadline,
+                rt.curr_epoch(),
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to deadline {} during its challenge window, \
+                        or the prior challenge window,
+                        or before {} epochs have passed since its last challenge window ended",
+                    params.to_deadline,
+                    policy.wpost_dispute_window
+                ));
+            }
```
Is this still required if we had done the verification of PoSt?

zhiqiangxu replied:
We only do the verification of from_deadline, but not to_deadline. 

It's obviously forbidden to change deadline partitions during deadline window and challenge window, not so sure about whether it's possible to allow modifications to the partition during the dispute window though. But the original `compact_partitions` function doesn't allow it during dispute window, so I followed its decision.

steven004 commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+        mut params: MovePartitionsParams,
+    ) -> Result<(), ActorError> {
+        let policy = rt.policy();
+        if params.from_deadline == params.to_deadline {
+            return Err(actor_error!(illegal_argument, "from_deadline == to_deadline"));
+        }
```
This could be deleted if we fail it when !deadline_available_for_move earlier (before the transaction handling).

zhiqiangxu replied:
same as reply for `deadline_available_for_move`

zhiqiangxu committed:
fix

hunjixin commented on actors/miner/src/lib.rs:
```diff
@@ -128,6 +128,7 @@ pub enum Method {
     ChangeBeneficiary = 30,
     GetBeneficiary = 31,
     ExtendSectorExpiration2 = 32,
+    MovePartitions = 33,
```
类似下面，用exported方法

zhiqiangxu replied:
`CompactPartitions` 也没用exported，这里的规范是什么？

hunjixin replied:
相当于hash一个方法数出来

hunjixin replied:
frc42规范是这样的

zhiqiangxu replied:
这个知道，我指的是什么情况需要用exported方法？因为CompactPartitions也没用。貌似exported方法的好处是可以从evm调，暂时应该还没这个需求。

hunjixin commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+        rt: &impl Runtime,
+        mut params: MovePartitionsParams,
+    ) -> Result<(), ActorError> {
+        let policy = rt.policy();
```
可以挪到第一个条件下面

zhiqiangxu replied:
Done.

hunjixin commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+            return Err(actor_error!(
+                illegal_argument,
+                "invalid deadline {}",
+                if params.from_deadline >= policy.wpost_period_deadlines {
```
直接返回from/to就行了，用户自行判断

zhiqiangxu replied:
Done.

hunjixin commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+                state.load_deadlines(store).map_err(|e| e.wrap("failed to load deadlines"))?;
+
+            // only try to do synchronous Window Post verification if from_deadline doesn't satisfy deadline_available_for_compaction
+            if !deadline_available_for_compaction(
```
建议把这些检查都封装到move里面，这里调一下move函数，函数有点太长了，deadline_available_for_compaction最后也抽到deadline_available_for_move里面，毕竟还是两个功能，现在可以套用，以后不一定可以套用。

zhiqiangxu replied:
主要是verify_windowed_post篇幅比较长，另外2处verify_windowed_post也是这个风格，先跟风下-_-b

zhiqiangxu replied:
这里两个check后续的处理不一样，所以没法合成一个判断。

zhiqiangxu replied:
还是说指的是to_deadline的deadline_available_for_compaction判断？那个倒是可以跟deadline_available_for_move合并，不过错误消息要改下

zhiqiangxu committed:
nit

hunjixin commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+                    )
+                })?;
+
+            if partitions.len() == 0 {
```
这个条件检查可以挪到前面么

hunjixin replied:
0等于全是吗

zhiqiangxu replied:
挪到`let mut from_deadline`和`let mut to_deadline`之间？感觉没啥必要

hunjixin replied:
可以上面定义的变量拿到这里吗，定义和使用贴近一点。

zhiqiangxu replied:
对的，0等于全

zhiqiangxu replied:
> 可以上面定义的变量拿到这里吗，定义和使用贴近一点。

这些变量下面也要用到。

hunjixin replied:
```
           let partitions = &mut params.partitions;
            if partitions.len() == 0 {
```

zhiqiangxu replied:
Done.

zhiqiangxu committed:
mod error

hunjixin commented on actors/miner/src/lib.rs:
```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+            }
+
+            let (live, dead, removed_power) =
+                from_deadline.remove_partitions(store, partitions, from_quant).map_err(|e| {
```
看remove的实现会处理expiration_epochs， 一个疑问是，如果check过windows和faulty power，这里还会有expiration_epochs集合么？可以稍微优化一下remove_partitions，如果该集合长度为0，可以跳过。

zhiqiangxu replied:
没get到，再具体一些，哪段？

zhiqiangxu committed:
nit

zhiqiangxu committed:
fmt

zhiqiangxu committed:
fix ci

zhiqiangxu committed:
fix bug

zhiqiangxu committed:
add test

zhiqiangxu committed:
add more test

zhiqiangxu committed:
Merge remote-tracking branch 'origin/master' into feature/move_partition_verify

zhiqiangxu committed:
partial fix for review

zhiqiangxu committed:
Merge remote-tracking branch 'origin/master' into feature/move_partition_verify

zhiqiangxu committed:
adjust test

zhiqiangxu committed:
use .context_code

zhiqiangxu committed:
fix for test

zhiqiangxu committed:
disallow empty partitions

zhiqiangxu committed:
refactor deadline_available_for_move

zhiqiangxu committed:
fix for clippy

zhiqiangxu committed:
minor opt

zhiqiangxu committed:
only verify_windowed_post once

zhiqiangxu committed:
mod error msg

zhiqiangxu committed:
1. verify_window_post batch by batch
2. move partitions intact, the `expirations_epochs` and `early_terminated` may have a gap of 24H

zhiqiangxu committed:
fix ci

zhiqiangxu committed:
mod check for epoch

zhiqiangxu committed:
partial review fix

zhiqiangxu committed:
Merge remote-tracking branch 'origin/master' into feature/move_partition_verify

zhiqiangxu committed:
adjust test

zhiqiangxu committed:
refactor with Partition::adjust_for_move

zhiqiangxu committed:
share the language with FIP

zhiqiangxu committed:
deadline_available_for_move => ensure_deadline_available_for_move

zhiqiangxu committed:
add some doc comment

zhiqiangxu committed:
more renaming

zhiqiangxu committed:
more renaming

zhiqiangxu committed:
Merge remote-tracking branch 'origin/master' into feature/move_partition_verify

zhiqiangxu committed:
rename + merge master

zhiqiangxu committed:
mod wording

zhiqiangxu committed:
fix test

zhiqiangxu committed:
renaming in test

zhiqiangxu committed:
apply alex's idea of not re-quantizing at all.

zhiqiangxu committed:
1. forbid moving when there're early terminations
2. remove `adjust_for_move`

zhiqiangxu committed:
rm anyhow::Ok

zhiqiangxu committed:
Merge branch 'master' into feature/move_partition_verify

zhiqiangxu committed:
minor optimization by observing that partition `faulty_power` should be zero when `faults` is empty

zhiqiangxu committed:
adjust find_sectors_by_expiration for not re-quantizing

zhiqiangxu committed:
add test

zhiqiangxu committed:
fix for review

zhiqiangxu committed:
add a comment about not re-quantizing when moving expirations_epochs

zhiqiangxu committed:
minor optimization

zhiqiangxu committed:
avoid scanning the same range twice

zhiqiangxu committed:
1. review fix
2. add check for `max_partitions_per_deadline`

zhiqiangxu committed:
fix comment

zhiqiangxu committed:
use with_context_code
